### PR TITLE
Show correct WooCommerce pages status if they are using blocks vs shortcode.

### LINKED ISSUE #3565: When Checkout and Cart blocks are used instead of checkout flow shortcodes - SSR report shows error for relevant pages.
The Woo Site Status Report does checks on whether shortcodes for Checkout and Cart and will show an error if the designated page for those does not contain the relevant shortcode. However, with the advent of the Checkout and Cart blocks, it's entirely feasible for some stores to be using the blocks instead of the shortcodes and this error could be alarming to store merchants. It could also contribute to confusion for Woo HEs troubleshooting sites using the SSR.

Here's an example where I've replace the checkout shortcode on my store checkout page with the Checkout block:

We should be able to tweak the status report to detect if the block exists and update the messaging appropriately. This might need updated in Woo Core, I haven't looked yet.

### PR DESCRIPTION
**Pull request by budzanowski**
This PR updates the WooCommerce status page to also consider WooCommerce Blocks for the Cart and Checkout pages status. Additionally, it displays the shortcode and or block slug in case they were not found ( it looks like this was the original intention some time ago bc the `sprintf` argument was there - just the placeholder was missing.

image

In the case when previously a page that had checkout block was used for checkout ( the same situation for cart ) the status page was showing that the page is missing the shortcode. Now the status is correct.

image

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes woocommerce/woocommerce-gutenberg-products-block/issues/3565 .

#### regression test
1. WC pages set up correctly, Cart and Checkout pages ( shortcode ) exist
2. Go to the WC status page and check that the pages are correctly detected

#### missing setup test
1. Select a page that has no shortcode or block for the Cart ( or Checkout ) page.
2. Go to the status view.
3. There should be an indication that the shortcode or the block is missing.

#### block present
1. Select a page that has the woocommerce/checkout block for the Checkout page.
2. Go to the status view.
3. The Checkout should be marked as green.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Check if Cart and Checkout pages are using block instead of shortcode for the Status check.

### CODE REVIEW

File: includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php
Comment (Aljullu) [inline includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:1234]: I feel a bit weird calling `WC_Tracker` for something which is not really related to the tracker.

Out of the scope for this PR, but at some point we might want to abstract that method somewhere outside the tracker.

**Review by Aljullu — DISMISSED** [review]
Nice to see this getting fixed! Code looks good to me and it tests well. :ok_hand:

File: includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php
Comment (budzanowski) [reply includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:1234]: Absolutely, it feels wired, but the use is so small that I did not want to abstract too early.

Comment (haszari) [reply includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:1234]: FYI the checks in `get_block_tracker_data` are a bit primitive, could be improved - https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/3224#issuecomment-707391821

There's a new improved approach using `parse_blocks()` in a PR that didn't merge, could be adapted into core: https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/bbef0a6763f3c5073902be248820badc09bc2496

It's might be worthwhile spending a bit of time tidying this up and abstracting out the detection logic. We don't often touch this code, so possibly worth investing a little more effort?

Comment (budzanowski) [reply includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:1234]: I don't mind doing it, cleanup makes sense. But I don't want to do it as a part of this PR. The scope should be separated. I don't mind postponing this and working on a cleaned-up version if this is what we think this is how this should be done.

Comment (budzanowski) [reply includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:1234]: Or merge as it is and follow up with a cleanup.

File: includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php
Comment (nerrad) [inline includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:1180]: Why is this (or for that matter shortcode too... but that's already a done deal) filtered? does it need to be filtered?

File: includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php
Comment (nerrad) [reply includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:1234]: A couple thoughts from my perspective here:

- I don't like using `WC_Tracker:: get_block_tracker_data` as this on the surface appears to be a violation of separation of concerns. For now the method does what you want, but it is a part of a class dedicated to WC_Tracker events so there's no guarantee it will do what you want in future refactors (or do what you want without any side-effects).
- To that point, extraction of whether blocks are in a page should not even be a method on `WC_Tracker` itself but rather something `WC_Tracker` consumes as a service.

I recognize the desire to keep scope from creeping in this pull and normally I'd agree, but this is severe enough that I'd like to avoid compounding the tech debt that already exists by pulling in `WC_Tracker` usage. I took a brief look at the code Rua shared and it should be trivial enough to create a new service class in Woo for explicitly solving this block detection problem (using the code Rua shared, or some variation of it). If you want to avoid creating an entire new class, you can add it as private methods in this class (`WC_REST_System_Status_V2_Controller`) for now and future follow ups could extract to a service class and implement in WC_Tracker too. I'd prefer creating the class.

**Review by nerrad — COMMENTED** [review]
Coming along great! Just adding my two cents.

File: includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php
Comment (budzanowski) [reply includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:1180]: This allows 3rd party plugins to provide valid blocks that can act as cart or checkout. If we don't filter it then `woocommerce/cart` becomes the only valid block here.

**Comment by budzanowski** [comment]
I have refactored the code in a way that the structure should now be more logical and maintainable. I have created a new utils class. I have moved some methods ( private ) from the tracker class there so we now have a central point for blocks related utils. I have also added a small and simple utility function that checks if a page contains a block.
@Aljullu @haszari @nerrad ?

File: includes/blocks/class-wc-blocks-utils.php
Comment (nerrad) [inline includes/blocks/class-wc-blocks-utils.php:69]: minor nit, grammar - should be `does_page_contain_block`. However, I also think I'd prefer something like this:

```suggestion
	/**
	 * Check if a given page contains a particular block.
	 *
	 * @param int|WP_Post $page   Page post ID or post object.
	 * @param string $block_name The name (id) of a block, e.g. `woocommerce/cart`.
	 * @return bool|null         Boolean value if the page contains the block or not. Null in case the page does not exist.
	 */
	public static function has_block_in_page( $page, $block_name ) {
```

I find following the convention of `has_` or `is_` prefixes for methods that are checks returning boolean to be really great to use for easier recall.

Also, I get why you're returning null here for when a page doesn't exist, but for the purpose of this method, I think it's best to not support that use case and mix concerns. It should only really be used to determine if the block exists in the page or not (if the page doesn't exist that's irrelevant?).

File: includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php
Comment (nerrad) [reply includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:1180]: Rationale sounds fair, I just don't think we should expose this filter until someone demonstrates why they'd want it. Not a blocker imo though, but probably should be weighed in on by @woocommerce/proton.

File: includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php
Comment (nerrad) [inline includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:1235]: Can probably just do something like:

```suggestion
				$block_present = WC_Block_Utils:does_page_contains_block( $page_id, $values['block'] );
```

**Review by nerrad — COMMENTED** [review]
Looks great Bartek! Thanks for handling the feedback. Probably should add some phpunit tests for the new utility class. Also probably should get the 👌 from someone on @woocommerce/proton before merging.

**Comment by haszari** [comment]
Looks good to me, nice work @budzanowski !

**Comment by budzanowski** [comment]
@nerrad Unit Tests are done. Now I need to find someone from Proton for a review :P

File: includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php
Comment (jonathansadowski) [inline includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:1180]: Thanks for the explanation @budzanowski, but I agree with @nerrad that we should hold off on this filter. Can we remove the filter before moving forward with this?

File: includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php
Comment (jonathansadowski) [inline includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:1185]: Same rationale to this filter. Thanks.

**Review by jonathansadowski — DISMISSED** [review]
Thanks for tackling this. Looking good to me. I agreed with the previous comment about the additional filters, but I think if we address that, this should be good to go.

File: includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php
Comment (budzanowski) [reply includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php]: Done.

File: includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php
Comment (budzanowski) [reply includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php]: Done.

**Comment by budzanowski** [comment]
@jonathansadowski review suggestions applied.

**Review by claudiosanches — APPROVED** [review]
Thanks!

**Pull request merged**
